text/0000-decorators.md

```diff
@@ -0,0 +1,1230 @@
+inconsistent, and could lead users to think that they are safe when they are
+not. This RFC suggests that we choose option 3 for this reason. When the actions
+namespace was introduced, lifecycle hooks like `destroy` and `click` were less
+commonly known and used (event listeners were not uncommon). Most Ember users
+know they exist now, and will be aware that implementing an action with the same
+name is not recommended. In addition, eslint rtules can be added to hint against
+these collisions.
+
+This also aligns with the future laid out by Glimmer components which will be
+discussed more thoroughly in upcoming RFCs, where actions will not need to be
+decorated at all, and will be plain methods:
+
+```js
+export default class ButtonComponent extends GlimmerComponent {
+  onClick() {
+    // handle click
+  }
+}
+```
+```hbs
+<button onclick={{action this.onClick}}>Click me!</button>
+```
+
+This pattern can be used today with legacy components and controllers, but
+routes still depend heavily on `sendAction` and actions being made available
+internally via the actions object, so for the time being, we need a way to mark
+methods as actions.
+
+#### Usage and API
+
+The API for this new decorator would be much simpler than the computed API,
+since it is only used as a decorator without parameters.
+
+```ts
+// Technically `action` is a function, but we can't type it transparently that way
+const action: MethodDecorator;
+```
+
+Attempting to pass any parameters to the decorator, or to apply the decorator to
+anything other than a class method, will throw an error.
+
+## How we teach this
+
+Teaching decorators is intrinsically tied to a wider shift in the Ember
+programming model - the Ember Octane edition. From a teaching perspective, this
+edition will be completely overhauling the guides and updating all of the best
+practices as they stand. New users should see native class syntax with
+decorators as the _default_, and should not ever have to write a legacy class
+or see an example for one.
+
+With this RFC, the majority of existing examples in the Ember guides will be
+updatable to native class syntax. The exception would be examples of legacy
+components, which would be addressed separately by updating the guides to
+Glimmer Components (proposed in a separate RFC). Otherwise, all examples in the
+guides should be updated.
+
+### Updating and Interop
+
+For existing users, or users who have to interact with legacy code from a modern
+context, it'll be important to have a reference for the legacy object model.
+The current section on the object model in the guides can be moved to a legacy
+section, and a section on updating should be added. Links to relevant codemods,
+such as the
+[ember-es6-class-codemod](https://github.com/scalvert/ember-es6-class-codemod),
+should be included.
+
+## Drawbacks
+
+* The `ComputedProperty` class has long been considered intimate API. Even with
```

Comment:
Is this true? We usually reserve the term "intimate API" for private APIs whose use is widespread enough that we cannot make certain changes without causing some painful disruption to the ecosystem.

`ComputedProperty` is public API https://www.emberjs.com/api/ember/release/classes/ComputedProperty

Reply:
Hoo boy, that makes things a bit tricky...

it seems like the documentation there is mainly for usage of `computed`, but yes I do think this constitutes public. I don't think this is very widely used, but we should look into the major users and make sure we have an update path. I do think it should still be possible to use prototypical inheritance, but `class` syntax to extend would be out.

Reply:
The Router type gives precedent for having a class that is public for use and naming the type in TS but *not* public for subclassing. That might be useful here?